### PR TITLE
Create a LazyString type, for places (like errors) which we would prefer to defer the actual work

### PR DESCRIPTION
**Pull request by Keno**
One common theme when looking at inference logs is that formatting error messages
tends to pull in the whole I/O and printing system which a) can take some time to
infer and b) adds backedges if that stuff gets invalidated. Both seem wasteful.
This is an attempt to try to improve the situation. For various error structs
that used to contain a string, they now take either a String or a closure that
will format the required string on demand when showing the error.

I have some tooling to count the total number of methods that inference looks
at and went through and switched over all the error messages that are used
by *(::Matrix{Float64}, ::Matrix{Float64} to the new setup. Doing this
reduces the total number of methods inference looks at by about 20%,
which seems like a worthwhile improvement.

Obviously there are a lot of error sites that could potentially throw, but this approach is backwards compatible, so if we like the direction, we can merge it and then we can gradually go through and clean up the callsite (or somebody could write a femtocleaner pass to do it).

### CODE REVIEW

**Comment by JeffBezanson** [comment]
This is not my favorite way to deal with this, since the large number of types generated for the closures is also pretty sub-optimal for the compiler. A possible alternative is to have a macro that makes string interpolation lazy, i.e. returning an object with the arguments. That might be useful for other purposes as well.

**Comment by Keno** [comment]
Alright, how do you like this version? Seems to still give the same benefits in inference (modulo a few extra LazyString constructors).

**Comment by Keno** [comment]
@JeffBezanson do you like this approach better? If so, I'll fix up this PR and get it ready.

**Comment by StefanKarpinski** [comment]
I kind of like this as an implementation, but it suggest to me that having a macro for throwing errors would be beneficial, e.g. `@throw DomainError "blah $x blah"`.

File: base/strings/lazy.jl
Comment (JeffBezanson) [inline base/strings/lazy.jl]:
```suggestion
function print(io::IO, s::LazyString)
```

File: base/strings/lazy.jl
Comment (c42f) [inline base/strings/lazy.jl]:
```suggestion
of functions).
```

**Comment by c42f** [comment]
@StefanKarpinski related to a possible `@throw`, I've sometimes thought it would be useful to be able to attach arbitrary key-value pairs to exceptions in much the same way we attach them to log records. I think people sometimes have a temptation to log-then-throw for adding context which is really not ideal.

(See also this crazy idea https://github.com/JuliaLang/julia/issues/7026#issuecomment-532946483 on how we might unify return codes with exceptions. It would make even more sense if combined with pattern matching of key value pairs.)

**Comment by StefanKarpinski** [comment]
I've been using the logging infrastructure for PkgServer and it's quite nice to just list key-value pairs for values that may be of interest when looking at logs. Doing something similar for errors would make sense to me as well.

File: base/strings/lazy.jl
Comment (simeonschaub) [inline base/strings/lazy.jl:23]:
```suggestion
        expr, idx = Meta.parseatom(string, idx; filename=string(__source__.file))
```
That way we also get correct source information.

**Comment by Keno** [comment]
I think this is good to go. Will merge in a bit so if there are any objections, say so now.

**Review by c42f — APPROVED** [review]
I think this is good useful utility, and I like the `@lazy_str` export. For minimalism I could live without the `LazyString` export, but leaving it in also seems harmless enough.

File: base/abstractarray.jl
Comment (Keno) [reply base/abstractarray.jl:917]: I just did the straightforward replacement. We can change these for consistency in a followup.

Comment (c42f) [reply base/abstractarray.jl:917]: Certainly fine by me.

**Comment by vtjnash** [comment]
This seems like it might potentially move in the opposite direction of the recent escape analysis-driven design changes, where we wanted to capture objects less frequently.

File: base/strings/lazy.jl
Comment (vtjnash) [inline base/strings/lazy.jl:13]: It might be better to allocate an Array here, so we do not need to create a whole new Tuple type and can examine and iterate it more easily later?

File: base/strings/lazy.jl
Comment (vtjnash) [inline base/strings/lazy.jl:12]: It feels a bit cheeky to declare that this is a kind of String, then implement none of the interface for one.

We could potentially make this object mutable, then swap the field for the string'd value when you first access it, perhaps?

File: base/strings/lazy.jl
Comment (Keno) [reply base/strings/lazy.jl:13]: It can be be an immutable array once that's merged, but currently the analysis in #43852 is not strong enough to model the mutation of the mutable array at the moment. I think for the current use cases, a tuple is mostly fine, since the types will be known at compile time and are stable. Plus it's the error path anyway, which we heavily pessimize and in particular, don't infer, so it'll allocate tons of tuple types anyway.

File: base/strings/lazy.jl
Comment (Keno) [reply base/strings/lazy.jl:12]: Yes, I think that's reasonable. Let me do that.

**Comment by Keno** [comment]
> This seems like it might potentially move in the opposite direction of the recent escape analysis-driven design changes, where we wanted to capture objects less frequently.

Well, as I said in those discussions, I don't like the idea to string summarize on error construction and if there escaping issues, we should insert maybecopies.

File: base/strings/lazy.jl
Comment (KristofferC) [inline base/strings/lazy.jl:47]: Is there any worry about race conditions for code like this?

Comment (vtjnash) [reply base/strings/lazy.jl:47]: Yes. We can use an atomic acquire and cmpswap release here to fix that. (Or perhaps create an "assign_once" type, that manages this pattern for us?)

**Comment by aviatesk** [comment]
> > This seems like it might potentially move in the opposite direction of the recent escape analysis-driven design changes, where we wanted to capture objects less frequently.
> 
> Well, as I said in those discussions, I don't like the idea to string summarize on error construction and if there are escaping issues, we should insert maybecopies.

I'm still not convinced that maybecopy is the right solution. One problem of the maybecopy idea is that it will require a global transformation, or many similar builtin function definitions.
E.g. if we want to allocate `a` on the stack in the following example, where a` can only escape via possible bounds error in `arrayref` call, we need to change `%24 = Base.arrayref(true, %2, 2)::Int64` to something like `%24 = Base.arrayref_copy_on_throw(true, %2, 2)::Int`, where `arrayref_copy_on_throw` uses another version of `BoundsError` constructor that makes a copy of `a`:
```julia
julia> const Lx = Ref{Any}();

julia> @noinline storelen!(xs) = Lx[] = length(xs);

julia> code_escapes() do
          a = [1,2,3]
          storelen!(a)
          a[2]
      end
#9() in Main at REPL[9]:2
2 ◌  1 ─ %1  = Core.tuple(1, 2, 3)::Core.Const((1, 2, 3))                                │╻      vect
  ↑  │   %2  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Vector{Int64}, svec(Any, Int64), 0, :(:ccall), Vector{Int64}, 3, 3))::Vector{Int64}
  ◌  └──       goto #7 if not true                                                       ││     
  ✓  2 ┄ %4  = φ (#1 => 1, #6 => %16)::Int64                                             ││     
  ✓  │   %5  = φ (#1 => 1, #6 => %17)::Int64                                             ││     
  ✓  │   %6  = φ (#1 => 1, #6 => %9)::Int64                                              ││     
  ◌  │   %7  = Base.getfield(%1, %4, true)::Int64                                        ││╻      getindex
  ◌  │         Base.arrayset(false, %2, %7, %6)::Vector{Int64}                           ││╻      setindex!
  ◌  │   %9  = Base.add_int(%6, 1)::Int64                                                ││     
  ◌  │   %10 = (%5 === 3)::Bool                                                          │││╻      ==
  ◌  └──       goto #4 if not %10                                                        │││    
  ◌  3 ─       Base.nothing::Nothing                                                     │││    
  ◌  └──       goto #5                                                                   │││    
  ◌  4 ─ %14 = Base.add_int(%5, 1)::Int64                                                │││╻      +
  ◌  └──       goto #5                                                                   ││╻      iterate
  ✓  5 ┄ %16 = φ (#4 => %14)::Int64                                                      ││     
  ✓  │   %17 = φ (#4 => %14)::Int64                                                      ││     
  ◌  │   %18 = φ (#3 => true, #4 => false)::Bool                                         ││     
  ◌  │   %19 = Base.not_int(%18)::Bool                                                   ││     
  ◌  └──       goto #7 if not %19                                                        ││     
  ◌  6 ─       goto #2                                                                   ││     
  ◌  7 ┄       goto #8                                                                   ││     
3 ◌  8 ─       invoke Main.storelen!(%2::Vector{Int64})::Any                             │      
4 ◌  │   %24 = Base.arrayref(true, %2, 2)::Int64                                         │╻      getindex
  ◌  └──       return %24                                                                │
```
This is because the constructor of `BoundsError` is opaque from this frame, and it's tricky to change the semantics of it based on the analysis on this frame. So probably the most sensible way to implement the maybecopy idea is to define "no-capturing on exception" alternatives of certain builtin functions whose no-throw checks inherently need to be conservative.

To me that sounds more complicated and I'd say it's much simpler if we change the semantics of certain error constructors so that they don't capture and let analyzers and optimizers assume that invariant.

**Comment by Keno** [comment]
Yes, the initial implementation of `maybecopy` is just to always copy the array, so you can assume no-escape. The issue people had with that is that the array may be big, so we want to have the flexibility to not do that in contexts where memory allocations are supposed to be tightly controlled. That's why the `maybecopy` is useful, because the optimizer can choose. I agree with you of course that it requires a global transformation, but I also think the extra memory issue is somewhat theoretical, and by the time it becomes a problem we should be able to do the compiler transform.

**Pull request merged**
